src/cascade/input_data/configuration/builder.py

```diff
@@ -26,6 +58,218 @@ def initial_context_from_epiviz(configuration):
+        if mul_cov_config.mulcov_type == "rate_value":
+            if target_dismod_name not in PRIMARY_INTEGRANDS_TO_RATES:
+                raise SettingsToModelError(f"Can only set a rate_value on a primary integrand. "
+                                           f"{mul_cov_config.measure_id} is not a primary integrand.")
+            add_to_rate = getattr(context.rates, PRIMARY_INTEGRANDS_TO_RATES[target_dismod_name])
+            add_to_rate.covariate_multipliers.append(covariate_multiplier)
+        else:
+            add_to_integrand = getattr(context.outputs.integrands, target_dismod_name)
+            if mul_cov_config.mulcov_type == "meas_value":
+                add_to_integrand.value_covariate_multipliers.append(covariate_multiplier)
+            elif mul_cov_config.mulcov_type == "meas_std":
+                add_to_integrand.std_covariate_multipliers.append(covariate_multiplier)
+            else:
+                raise RuntimeError(f"mulcov_type isn't among the three {configuration.mulcov_type}")
+
+
+def reference_value_for_covariate_mean_all_values(cov_df):
+    """
+    Strategy for choosing reference value for country covariate.
+    This one takes the mean of all incoming covariate values.
+    """
+    return float(cov_df["mean_value"].mean())
+
+
+def covariate_to_measurements_nearest_favoring_same_year(measurements, covariates):
+    """
+    Given a covariate that might not cover all of the age and time range
+    of the measurements select a covariate value for each measurement.
+    This version chooses the covariate value whose mean age and time
+    is closest to the mean age and time of the measurement in the same
+    year. If that isn't found, it picks the covariate that is closest
+    in age and time in the nearest year. In the case of a tie for distance,
+    it averages.
+
+    Args:
+        measurements (pd.DataFrame):
+            Columns include ``age_lower``, ``age_upper``, ``time_lower``,
+            ``time_upper``. All others are ignored.
+        covariates (pd.DataFrame):
+            Columns include ``age_lower``, ``age_upper``, ``time_lower``,
+            ``time_upper``, and ``value``.
+
+    Returns:
+        pd.Series: One row for every row in the measurements.
+    """
+    # Rescaling the age by 120 means that the nearest age within the year
+    # will always be closer than the nearest time across a full year.
+    tree = spatial.KDTree(list(zip(
+        covariates[["age_lower", "age_upper"]].mean(axis=1) / 240,
+        covariates[["time_lower", "time_upper"]].mean(axis=1),
+        covariates["x_sex"]
+    )))
+    _, indices = tree.query(list(zip(
+        measurements[["age_lower", "age_upper"]].mean(axis=1) / 240,
+        measurements[["time_lower", "time_upper"]].mean(axis=1),
+        measurements["x_sex"],
+    )))
+    return pd.Series(covariates.iloc[indices]["mean_value"].values, index=measurements.index)
+
+
+def convert_age_year_ids_to_ranges(with_ids_df, age_groups_df):
```

Comment:
I first thought of this as "age year" rather than as "age and year", but doesn't seem like a big deal, though looks like it's "age sex and year" business.